File: 03-SOLID-Principles/src/go/lsp.go
package main

import "fmt"

// LSP — Liskov Substitution Principle
// Separate interfaces for separate capabilities

type BirdLike interface {
	Name() string
}

type Flyable interface {
	Fly()
}

type Sparrow struct{}
func (Sparrow) Name() string { return "Sparrow" }
func (Sparrow) Fly()         { fmt.Println("Sparrow is flying!") }

type PenguinGood struct{}
func (PenguinGood) Name() string { return "Penguin" }
func (PenguinGood) Swim()        { fmt.Println("Penguin is swimming!") }
// No Fly() — Penguin doesn't satisfy Flyable

func MakeFly(f Flyable) {
	f.Fly()
}

File: 03-SOLID-Principles/src/rust/lsp.rs
/// LSP — Liskov Substitution Principle
/// Separate traits for separate capabilities

pub trait BirdLike {
    fn name(&self) -> &str;
}

pub trait Flyable {
    fn fly(&self);
}

pub struct Sparrow;
impl BirdLike for Sparrow {
    fn name(&self) -> &str { "Sparrow" }
}
impl Flyable for Sparrow {
    fn fly(&self) { println!("Sparrow is flying!"); }
}

pub struct Penguin;
impl BirdLike for Penguin {
    fn name(&self) -> &str { "Penguin" }
}
// Penguin does NOT implement Flyable — compiler enforces LSP
impl Penguin {
    pub fn swim(&self) { println!("Penguin is swimming!"); }
}

pub fn make_fly(f: &dyn Flyable) {
    f.fly();
}
